### PR TITLE
abstract_apply: Don't drop effects of `iterate`'d calls

### PR DESCRIPTION
**Pull request by Keno**
We were accidentally dropping the effects of calls from `iterate` calls performed during abstract_iteration. This allowed calls that were not actually eligible for (semi-)concrete evaluation to go through that path anyway. This could cause incorrect results (see test), though it was usually fine, since iterate call tend to not have side effects. It was noticed however in #47688, because it forced irinterp down a path that was not meant to be reachable (resulting in a TODO error message). For good measure, let's also address this todo (since it is reachable by external absint if they want), but the missing effect propagation was the more serious bug here.

### CODE REVIEW

File: base/compiler/stmtinfo.jl [hunk 1]
```diff
@@ -114,6 +114,7 @@ Each (abstract) call to `iterate`, corresponds to one entry in `ainfo.each::Vect
 """
 struct AbstractIterationInfo
     each::Vector{CallMeta}
+    complete::Bool
```
Comment (aviatesk) [inline base/compiler/stmtinfo.jl:117]: Does this change is necessary for this PR or it is a separate fix on the inlining algorithm?

File: base/compiler/abstractinterpretation.jl [hunk 2]
```diff
@@ -1482,7 +1492,7 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
                 # ... but cannot terminate
                 if !may_have_terminated
                     #  ... and cannot have terminated prior to this loop
-                    return Any[Bottom], nothing
+                    return AbstractIterationResult(Any[Bottom], AbstractIterationInfo(calls, false), EFFECTS_UNKNOWN′)
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:1495]: `EFFECTS_THROWS` seems to be more consistent with the change on L1465 above?
```suggestion
                    return AbstractIterationResult(Any[Bottom], AbstractIterationInfo(calls, false), EFFECTS_THROWS)
```

File: base/compiler/abstractinterpretation.jl [hunk 3]
```diff
@@ -1548,6 +1556,12 @@ function abstract_apply(interp::AbstractInterpreter, argtypes::Vector{Any}, si::
                 end
                 cti = Any[Vararg{argt}]
             end
+            effects = merge_effects(effects, ai_effects)
+            if info !== nothing
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:1560]:
```suggestion
            # merge effects of the `iterate` call
            effects = merge_effects(effects, ai_effects)
            # merge effects of call(s) with the iterated arguments
            if info !== nothing
```

File: test/compiler/inference.jl [hunk 4]
```diff
@@ -4633,3 +4633,19 @@ end |> only === Type{Float64}
...
+	n <= N ? (n, n+1) : nothing
+end
+foo47688() = tuple(CountsIterate47688{5}()...)
+bar47688() = foo47688()
+@test only(Base.return_types(bar47688)) == NTuple{6, Int}
+@test it_count47688 == 0
+@test isa(bar47688(), NTuple{6, Int})
+@test it_count47688 == 7
+@test isa(foo47688(), NTuple{6, Int})
+@test it_count47688 == 14
```
Comment (aviatesk) [inline test/compiler/inference.jl:4651]:
```suggestion
@test it_count47688 == 14

```

File: base/compiler/stmtinfo.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/stmtinfo.jl:117]: This PR adds all the calls into the `.each` field (partly to look at the affects, but mostly for Cthulhu's benefit). Previously the inlining algorithm used the presence or absence of this info to determine whether or not we had the complete set of iteration items. Now we need to keep track of it separately.

File: base/compiler/abstractinterpretation.jl (hunk 2 above)
Comment (Keno) [reply base/compiler/abstractinterpretation.jl:1495]: At this point, we have proven that we cannot terminate, but I'm not sure we've proven that the state has fixpointed, meaning that our call effects cover the largest possible effect set the call could produce during its execution.

File: base/compiler/abstractinterpretation.jl (hunk 3 above)
Comment (Keno) [reply base/compiler/abstractinterpretation.jl:1560]: It's not really the effects of the iterate call. It's more the effects of imprecision of the iterate call. The `info.each`, are all methods of `iterate`.

Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl:1560]: Ah, I understand.

**Review by aviatesk — APPROVED** [review]

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47846#issuecomment-1343966866) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/2934433_vs_6bfc6ac/report.md).

**Pull request merged**